### PR TITLE
Fix channel iteration

### LINKED ISSUE #307: Fix channel iteration methods
They must comply with the spec:
```
{{ iterate() -> Iterator<ChannelType> // RSN2, RTS2 }}

```
Instead, we have these:

```
{{func (ch *RealtimeChannels) All() []*RealtimeChannelfunc (c *RESTChannels) Range(fn func(name string, channel *RESTChannel) bool) {}}

```

### PR DESCRIPTION
**Pull request by sacOO7**
- Added `Iterator()` to `RealtimeChannels` and `RESTChannels`
- Refactored code/tests, removed default channel name sort and range from restchannels
- Internal discussion [link](https://ably-real-time.slack.com/archives/C0114KK7S6R/p1620302233050500)
- Fixes #307

### CODE REVIEW

File: ably/export_test.go [hunk 1]
```diff
@@ -147,5 +147,5 @@ type Result = result
 func (c *RESTChannels) Len() int {
```
Comment (tcard) [inline ably/export_test.go:147]: Not necessarily as part of this PR, but can we unexport or remove this? It isn't part of the spec, and arguably it's not very useful.

File: ably/realtime_channel.go [hunk 2]
```diff
@@ -85,26 +85,23 @@ func (ch *RealtimeChannels) Get(name string, options ...ChannelOption) *Realtime
...
 		chans = append(chans, c)
 	}
 	ch.mtx.Unlock()
-	chanSlice(chans).Sort()
 	return chans
 }
 
 // Exists returns true if the channel by the given name exists.
-func (c *RealtimeChannels) Exists(name string) bool {
+func (c *RealtimeChannels) Exists(name string) bool { // RSN2, RTS2
```
Comment (tcard) [inline ably/realtime_channel.go:104]: Same as above, this isn't part of the spec.

**Review by tcard — APPROVED** [review]
Looks good, but we could to make the whole Channels API spec-compliant by removing the extra methods.

File: ably/realtime_channel.go (hunk 2 above)
Comment (tcard) [reply ably/realtime_channel.go:104]: > Same as above, this isn't part of the spec.

Well, this clearly isn't true. Sorry.

File: ably/export_test.go (hunk 1 above)
Comment (sacOO7) [reply ably/export_test.go]: Removed

**Review by tcard — APPROVED** [review]

**Pull request merged**